src/game-mechanics.js

```diff
@@ -1,59 +1,89 @@
+    undefined
+  );
 
-export const validateWin = (gameState) => {
-  return winStates.filter((winState) => {
-    console.log(
-      gameState,
-      gameState[winState[0][0]][winState[0][1]] ===
-        gameState[winState[1][0]][winState[1][1]] &&
-        gameState[winState[1][0]][winState[1][1]] ===
-          gameState[winState[2][0]][winState[2][1]]
-    );
+const checkPlayerState = (boardSize) => ({ column, row }) =>
+  letIn({ eqBoardSize: eq(boardSize), gt0: gt(0) }, ({ eqBoardSize, gt0 }) =>
+    letIn(
+      {
+        columnStraight: column.some(eqBoardSize),
+        rowStraight: row.some(eqBoardSize),
+        diagonalStraight: column.all(gt0) && row.all(gt0),
+      },
+      (columnStraight, rowStraight, diagonalStraight) =>
+        columnStraight || rowStraight || diagonalStraight
+    )
+  );
 
-    return (
-      " " !== gameState[winState[0][0]][winState[0][1]] &&
-      gameState[winState[0][0]][winState[0][1]] ===
-        gameState[winState[1][0]][winState[1][1]] &&
-      gameState[winState[1][0]][winState[1][1]] ===
-        gameState[winState[2][0]][winState[2][1]]
-    );
-  });
-};
+/* Player State looks something like this, but can accept any string as a name
+ *
+ * {
+ *  x: { column: [0, 0, 0]
+ *     , row: [0, 0, 0]
+ *     }
+ *  o: { column: [0, 0, 0]
+ *     , row: [0, 0, 0]
+ *     }
+ * }
+ */
+const boardToPlayerState = (board) =>
+  board.reduce(
+    (playerState, row, rowIndex) =>
+      attributeSquareToPlayer(row)(rowIndex)(playerState),
+    {}
+  );
 
-export const hasWon = (gameState) => {
-  return validateWin(gameState).reduce(
-    (result, winState) => result || winState,
-    false
+const attributeSquareToPlayer = (row) => (rowIndex) => (acc) =>
+  row.reduce(
+    (acc, player, columnIndex) =>
+      modifyObject(acc)(player)((playerValue) =>
+        playerValue === undefined
+          ? {
+              columns: modifyArray(
+                columnIndex,
+                add(1),
+                makeArrayOf(row.length - 1)(0)
+              ),
+              rows: modifyArray(
+                rowIndex,
+                add(1),
+                makeArrayOf(row.length - 1)(0)
+              ),
+            }
+          : {
+              columns: modifyArray(columnIndex, add(1), playerValue.column),
+              rows: modifyArray(rowIndex, add(1), playerValue.row),
+            }
+      ),
+    acc
   );
-};
+
+const makeArrayOf = (size) => (contents) =>
+  Array.from(Array(size), constt(contents));
+const add = (a) => (b) => a + b;
+const gt = (a) => (b) => a > b;
+const eq = (a) => (b) => a === b;
+// NOTE: all of these are in my helper library
+const constt = (a) => () => a;
+const modifyObject = (obj) => (key) => (fn) => ({
+  ...obj,
+  [key]: fn(obj[key]),
+});
+const modifyArray = (idx, fn, arr) =>
+  arr.slice(0, idx) + fn(arr[idx]) + arr.slice(idx + 1);
+const compose = (f) => (g) => (x) => f(g(x));
+const letIn = (expr, functionBody) => functionBody(expr);
```

Comment:
It's good that `letIn` commit to expressions, but for javascript having a const in a block and it seems to serve the same purpose. It's also easier to see for a person who is just learning if they read it top to bottom, rather than as something that pops the first in the expression of the second.

Reply:
I agree, that it makes it a bit harder to read at first, but I think that it is also a nice way to get comfortable with higher order functions.